spec/std/openssl/ssl/hostname_validation_spec.cr

```diff
@@ -0,0 +1,54 @@
+require "spec"
+require "openssl/ssl/hostname_validation"
+
+describe OpenSSL::SSL::HostnameValidation do
+  describe "matches_hostname?" do
+    it "matches ipv4, ipv6" do
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("192.168.1.1", "192.168.1.1").should be_true
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("192.168.1.2", "192.168.1.1").should be_false
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("::1", "::1").should be_true
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("::1", "::2").should be_false
+    end
+
+    it "won't wildcard match against ip" do
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("*.168.1.1", "192.168.1.1").should be_false
+    end
+
+    it "skips trailing dot" do
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("example.com.", "example.com").should be_true
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("example.com", "example.com.").should be_true
+      OpenSSL::SSL::HostnameValidation.matches_hostname?(".example.com", "example.com").should be_false
+      OpenSSL::SSL::HostnameValidation.matches_hostname?("example.com", ".example.com").should be_false
```

Comment:
Found it again, it's in `X509_check_host(3)`:

> If set, X509_CHECK_FLAG_SINGLE_LABEL_SUBDOMAINS restricts name values which start with ".", that would otherwise match any sub-domain in the peer certificate,
       to only match direct child sub-domains.  Thus, for instance, with this flag set a name of ".example.com" would match a peer certificate with a DNS name of
       "www.example.com", but would not match a peer certificate with a DNS name of "www.sub.example.com"; this flag only applies to X509_check_host.